smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/CommandGenerator.java

```diff
@@ -0,0 +1,234 @@
+                // Add serialization and deserialization plugins.
+                writer.write("this.use(serializerPlugin(configuration, this.serialize));");
+                writer.write("this.use(deserializerPlugin<$L>(configuration, this.deserialize));", outputType);
+
+                // Add customizations.
+                addCommandSpecificPlugins();
+
+                // Resolve the middleware stack.
+                writer.write("\nconst stack = clientStack.concat(this.middlewareStack);\n");
+                writer.openBlock("const handlerExecutionContext: $$types.HandlerExecutionContext = {", "}", () -> {
+                    writer.write("logger: {} as any,");
+                });
+                writer.write("const { httpHandler } = configuration;");
+                writer.openBlock("return stack.resolve(", ");", () -> {
+                    writer.write("(request: $$types.FinalizeHandlerArguments<any>) => ");
+                    writer.write("  httpHandler.handle(request.request as $T, options || {}),",
+                                 applicationProtocol.getRequestType());
+                    writer.write("handlerExecutionContext");
+                });
+            });
+
+            writeSerde();
+
+            // Hook for adding more methods to the command.
+            writer.write("// Start section: $L", COMMAND_BODY_EXTRA_SECTION)
+                    .pushState(COMMAND_BODY_EXTRA_SECTION)
+                    .popState()
+                    .write("// End section: $L", COMMAND_BODY_EXTRA_SECTION);
+        });
+    }
+
+    private void addInputAndOutputTypes() {
+        writeInputOrOutputType(outputType, operationIndex.getInput(operation).orElse(null));
+        writeInputOrOutputType(inputType, operationIndex.getOutput(operation).orElse(null));
+        writer.write("");
+    }
+
+    private void writeInputOrOutputType(String typeName, StructureShape struct) {
+        // If the input or output are non-existent, then use an empty object.
+        if (struct == null) {
+            writer.write("export type $L = {};", typeName);
+        } else {
+            writer.write("export type $L = $T;", typeName, symbolProvider.toSymbol(struct));
+        }
+    }
+
+    private void addCommandSpecificPlugins() {
+        // Some plugins might only apply to specific commands. They are added to the
+        // command's middleware stack here. Plugins that apply to all commands are
+        // applied automatically when the Command's middleware stack is copied from
+        // the service's middleware stack.
+        for (RuntimeClientPlugin plugin : runtimePlugins) {
+            if (plugin.hasMiddleware() && plugin.getOperationNames().contains(operation.getId().getName())) {
+                writer.write("this.use($T.getMiddleware(configuration));", plugin.getSymbol());
+            }
+        }
+    }
+
+    private void writeSerde() {
+        writer.write("")
```

Comment:
Nice! So we treat serde plugin specially, so I don't need to worry about serde plugin takes serializer/deserializer as parameter in addition to config

Reply:
Yeah, serde is handled differently than plugins. This is because (de)serialization is heavily influenced by code-gen-time configuration settings. We've decided that we'll support a single application level protocol at any time for a client (though you could support multiple protocols at the same time if they use the same application protocol).